### DISCUSSION TITLE
Cell neighbor lists

### DISCUSSION THREAD

**Discussion by markmbaum** [nschloe/meshzoo#86] [Ideas]
It would be useful to provide lists of neighbor cells for each cell in a mesh. I don't know of the most efficient way to do it, but most applications require neighborhoods.

**Comment by nschloe** [comment]
That'd rather be the job of a mesh processing lib like [meshplex](https://github.com/nschloe/meshplex/).

_Edit:_ That being said, I don't think that cell neighbors is an efficient data structure. For triangles, you'd have to keep around four arrays: One for the triangles without neighbors, one for the triangles with one neighbor, etc. Plus a list that maps the triangle indices into each of the lists. Granted, most triangles will "probably" have three neighbors, but still.

In many applications, you can probably get away with just facet data information (basically a list of neighborhood info, like `[(1, 2), (1, 3), (2, 5), (2, 7), ...]`).

**Reply by markmbaum** [reply]
That's true, you may have to generate a list of arrays for some meshes where cells have different numbers of neighbors. The spherical (surface) mesh is closed, though, so you can have an N x 3 array of indices where row `i` indicates the neighbors of cell `i`.

**Reply by nschloe** [reply]
Out of curiosity, what's your application?

**Reply by markmbaum** [reply]
I'm not working on anything right now, but a few years ago I was doing some finite volume modeling on the surface of a sphere and on disks. Meshing was by far the most difficult part of coding. I would have definitely have used `meshzoo` to generate the meshes, but would still have needed to find all the neighborhoods.

**Reply by nschloe** [reply]
Why do you need neighbors for FEM?

**Reply by markmbaum** [reply]
not FEM, FVM

**Reply by nschloe** [reply]
Right. Same question though!

**Reply by markmbaum** [reply]
Would be nice to get model-ready meshes for simple domains like that without having to touch a fully featured mesh generator, but just a suggestion, of course!

**Reply by markmbaum** [reply]
> Right. Same question though!

Well, you need to compute various relations between cells and their neighbors, so you need to know what the neighborhoods are.

**Reply by markmbaum** [reply]
You're right that facets are enough, though

**Reply by nschloe** [reply]
For FVM you don't need neighbor info. It's sufficient to consider cells individually. The complicated thing is to split up each cell into the parts that "belong" the each of its nodes. Some years ago I wrote [meshplex](https://github.com/nschloe/meshplex) and [pyfvm](https://github.com/nschloe/pyfvm) to to just that.
